presto-docs/src/main/sphinx/connector/kudu.rst

```diff
@@ -0,0 +1,484 @@
+==============
+Kudu Connector
+==============
+
+The Kudu connector allows querying, inserting and deleting data in `Apache Kudu`_
+
+.. _Apache Kudu: https://kudu.apache.org/
+
+
+.. contents::
+    :local:
+    :backlinks: none
+    :depth: 1
+
+
+Compatibility
+-------------
+
+Connector is compatible with all Apache Kudu versions starting from 1.0.
+
+If the connector uses features that are not available on the target server, an error will be returned.
+Apache Kudu 1.7.0 is currently used for testing.
+
+
+Configuration
+-------------
+
+To configure the Kudu connector, create a catalog properties file
+``etc/catalog/kudu.properties`` with the following contents,
+replacing the properties as appropriate:
+
+  .. code-block:: none
+
+       connector.name=kudu
+
+       ## List of Kudu master addresses, at least one is needed (comma separated)
+       ## Supported formats: example.com, example.com:7051, 192.0.2.1, 192.0.2.1:7051,
+       ##                    [2001:db8::1], [2001:db8::1]:7051, 2001:db8::1
+       kudu.client.master-addresses=localhost
+
+       ## Optional restriction of tablets for specific tenant.
+       ## If a tenant is set, only Kudu tablets starting with `<tenant>.` will
+       ## be visible in Presto
+       #kudu.session.tenant=mytenant
+
+       #######################
+       ### Advanced Kudu Java client configuration
+       #######################
+
+       ## Default timeout used for administrative operations (e.g. createTable, deleteTable, etc.)
+       #kudu.client.defaultAdminOperationTimeout = 30s
+
+       ## Default timeout used for user operations
+       #kudu.client.defaultOperationTimeout = 30s
+
+       ## Default timeout to use when waiting on data from a socket
+       #kudu.client.defaultSocketReadTimeout = 10s
+
+       ## Disable Kudu client's collection of statistics.
+       #kudu.client.disableStatistics = false
+
+
+Querying Data
+-------------
+
+A Kudu table named ``mytable`` is available in Presto as table
+``kudu.default.mytable``. A Kudu table containing a dot is considered as
+a schema/table combination, e.g. ``dev.mytable`` is mapped to the Presto
```

Comment:
Yes, a Kudu table can contain multiple `.`. 
With the current mapping, a Kudu table like `part1.part2.part3` belongs to schema `part1` and is used in Presto as `select * from kudu.part1."part2.part3"`. As Presto supports schemas it would be a shame not to make use of it.

Reply:
@MartinWeindel  any plans to support schemas in Kudu? https://community.cloudera.com/t5/Interactive-Short-cycle-SQL/Exception-on-Creating-Kudu-table-mapping-in-Impala/m-p/61582/highlight/true#M3805 warns against such future development, which could later force us to change mapping of Kudu table `part1.part2.part3` to `kudu.default."part1.part2.part3"`.

Reply:
I've got very helpful feedback from William Berkely (Cloudera):

> I don't see any conflicts between that and any current or near-term work I know of happening in Kudu. There are a couple of related JIRAs for database support: [KUDU-2063](https://issues.apache.org/jira/browse/KUDU-2063) and [KUDU-2362](https://issues.apache.org/jira/browse/KUDU-2362).

> In fact, Impala does something similar right now. Kudu tables managed by Impala are named like impala::database_name.table_name. It might also be a good idea to add some kind of identifier like the initial "impala::" for presto-managed or -created tables, since many other integrations may use a database_name.table_name convention.

My conclusions:
1. Some dormant JIRA issues are existing in Kudu, but nothing will really change in the near future
2. Other integrations like Impala are going a similar way
3. I propose to add an optional prefix "presto::" for Presto-managed tables, which would be enabled by default. E.g. a table named `kudu.part1.part2` in Presto will be named as `presto::part1.part2` in Kudu.

@kokosing Do you agree, or what are your thoughts?

Reply:
I'm not sure how this answers the problem of mapping to a schema name on Presto side, what's the plan here.
Mapping Kudu's `part1.part2` to `"part1"."part2"` in Presto (i'm using `"` to avoid ambiguities) sounds cool & natural.
But then, you need to do something about:
- `part1.part2.part3` → `"part1"."part2.part3"` -- this is no longer so natural
- `part1` → `"default"."part1"` -- which looks like a special case in this context
- `part1..part2` → ?
- `.part1` → ?
- `part1...` → ?
- `...` → ?

While it's probably possible to come up with a coherent answer to all this questions, I think it's better to just have one simple answer -- no schemas in Kudu means no schemas on Presto side.

Also, from completely different angle, some tools tend to lock users into one chosen catalog/schema in Presto. (For example `pyhive` client will lock you in `hive.default`  by [default](https://github.com/dropbox/PyHive/blob/master/pyhive/presto.py#L82)). Scattering tables across schemas can be "counter useful" to users.


> I propose to add an optional prefix "presto::" for Presto-managed tables, which would be enabled by default. E.g. a table named kudu.part1.part2 in Presto will be named as presto::part1.part2 in Kudu.

This i don't understand. It sounds contrary to how other storage systems behave.
When John creates a table tab in Postgres, it's named "tab", not "john::tab" nor "john's tab".
When John creates a table tab in Hive via Hive/Presto/Spark/impala/YouNameIt, it's named "tab" again.

Then again, if Presto-created table "tab" ends up in Kudu as "presto::tab", then -- on read -- it needs to be stripped of this prefix, and be presented back  as "tab". That's hopefully obvious. But then, what should happen to tables "impala::tab", "presto::impala::tab", etc.? Should "impala::tab" be concealed from Presto? Should "impala::tab" end up having the same name as "presto::impala::tab" on Presto side?

It's giving me a hard time to think about consequences of this.

Reply:
I agree with all what Piotr just said.

Does the prefix `prestod::` solve any existing problem that could not be solved without it? 

Notice we do not use anything like that in other connectors, so it would be nice to have coherent behavior.

Also, I think that as a user I could find it awkward if Kudu connector had schemas that do not exists in Kudu backend system.

Reply:
It seems I have not explained the idea with the `presto::` prefix in sufficient depth.
- All tables without the `presto::`prefix will live in the `default` schema, e.g.
  - `table1` -> `kudu.default.table1`
  - `part1.part2` -> `kudu.default."part1.part2"`
  - `part1.part2.part3` -> `kudu.default."part1.part2.part3"`
  - `impala::part1.part2` -> `kudu.default."impala::part1.part2"`

If a Presto user creates a table in another schema, the table name will start with `presto::`, e.g.
  - `CREATE TABLE kudu.project1.table1 ...` => kudu table name `presto::project1.table1`
  - `CREATE TABLE kudu.part1."part2.part3" ...` => kudu table name `presto::part1.part2.part3`

but:
  - `CREATE TABLE kudu.default.table2 ...` => kudu table name `table2`
  - `CREATE TABLE kudu.default."part1.part2" ...` => kudu table name `part1.part2`

Some corner cases would need to be forbidden by the presto-kudu connector:
  - schema names with dots
  - tables named `kudu.default."presto::..."`

It would also be quite easy to make this feature optional with a connector property. Then the owner of the Presto installation can decide himself, if schemas can be simulated by the `presto::`/dot convention, i.e. "use on own risk"

Regarding the user perspective:
- There are a few users already using the connector. Nobody has complained about this feature.
- I'm not only a developer but also a user of this connector. And I need schemas in my projects. If you decide to drop this feature, I will need to maintain a fork with schema support.

Reply:
> It would also be quite easy to make this feature optional with a connector property.

Well, it looks a bit like it's a naming convention being set in stone here. If this is Kudu-global convention, then it's fine. If this is something organization-specific, then  -- if we really do need to support the convention -- it should have a toggle and be off by default.

> Some corner cases would need to be forbidden by the presto-kudu connector: [...] tables named `kudu.default."presto::..."`

What happens if a kudu table `presto::default.xxx` is created in Kudu (possibly by some external means, since you want the connector to reject that)?

Also, if user is allowed to create tables in schemas other than `default`, then they must also be allowed to create schemas. How schemas are represented in Kudu? I.e. what's the effect of `CREATE SCHEMA kudu.sss` in Presto?

Reply:
> So is this the comprimise: Have a toogle in the connector properties and turn it off by default?

+1. @kokosing ?
You can call the toggle eg  `kudu.emulate-schemas`

> If someone creates a kudu table presto::default.xxx externally, it will not be accessible with the connector.

Of course only when `kudu.infer-schema-name-from-table-name = true`.
Technically, you could expose such table as `kudu.default."presto::default.xxx"` (this would be unambiguous, since you map Presto tables `kudu.default.X` to `X` in Kudu). However, I don't mind if it remains unsupported.

> The connector maintains a special table in Kudu to persist the schemas.  [...] I suggest to rename it to `presto::$schemas`.

Nice.

Reply:
I am ok with this behaviour if it is hidden by a toggle.